packages/xo-server/src/xo-mixins/metadata-backups.js

```diff
@@ -0,0 +1,261 @@
+    const remoteIds = unboxIdsFromPattern(job.remotes)
+    if (remoteIds.length === 0) {
+      throw new Error('metadata backup job cannot run without remotes')
+    }
+
+    const poolIds = unboxIdsFromPattern(job.pools)
+    const isEmptyPools = poolIds.length === 0
+    if (!job.xoMetadata && isEmptyPools) {
+      throw new Error('no metadata mode found')
+    }
+
+    const app = this._app
+    const { retentionXoMetadata, retentionPoolMetadata } =
+      job?.settings[schedule.id] || {}
+
+    const scheduleDir = `${METADATA_BACKUP_DIR}/${schedule.id}`
+    const timestamp = safeDateFormat(Date.now())
+    const files = []
+    if (job.xoMetadata && retentionXoMetadata > 0) {
+      const xoDir = `${scheduleDir}/xo`
+      const dir = `${xoDir}/${timestamp}`
+
+      const data = JSON.stringify(await app.exportConfig(), null, 2)
+      const fileName = `${dir}/data.json`
+
+      const metadata = JSON.stringify(
+        {
+          jobId: job.id,
+          scheduleId: schedule.id,
+          timestamp,
+        },
+        null,
+        2
+      )
+      const metaDataFileName = `${dir}/metadata.json`
+
+      files.push({
+        executeBackup: defer(($defer, handler) => {
+          $defer.onFailure(() => handler.rmtree(dir))
+          return Promise.all([
+            handler.outputFile(fileName, data),
+            handler.outputFile(metaDataFileName, metadata),
+          ])
+        }),
+        dir: xoDir,
+        retention: retentionXoMetadata,
+      })
+    }
+    if (!isEmptyPools && retentionPoolMetadata > 0) {
+      files.push(
+        ...(await Promise.all(
+          poolIds.map(async id => {
+            const metadataDir = `${scheduleDir}/${id}`
+            const dir = `${metadataDir}/${timestamp}`
+
+            // TODO: export the metadata only once then split the stream between remotes
+            const stream = await app.getXapi(id).exportPoolMetadata(cancelToken)
+            const fileName = `${dir}/data`
+
+            const metadata = JSON.stringify(
+              {
+                jobId: job.id,
+                scheduleId: schedule.id,
+                poolId: id,
+                timestamp,
+              },
+              null,
+              2
+            )
+            const metaDataFileName = `${dir}/metadata.json`
+
+            return {
+              executeBackup: defer(($defer, handler) => {
+                $defer.onFailure(() => handler.rmtree(dir))
+                return Promise.all([
+                  (async () => {
+                    const outputStream = await handler.createOutputStream(
+                      fileName
+                    )
+                    $defer.onFailure(() => outputStream.destroy())
+
+                    // 'readable-stream/pipeline' not call the callback when an error throws
+                    // from the readable stream
+                    stream.pipe(outputStream)
+                    return pFinished(stream)
```

Comment:
Nope, you don't want to wait for the metadata export to be finished but the output file to be completely written.